packages/devtools_app/lib/src/shared/memory/classes.dart

```diff
@@ -0,0 +1,235 @@
+  final objectsExcludedFromRetainedSize = <int>{};
+
+  @override
+  bool get isZero => objects.isEmpty;
+
+  @override
+  void countInstance(
+    HeapSnapshotGraph graph,
+    int index,
+    List<int>? retainedSizes, {
+    required bool excludeFromRetained,
+  }) {
+    super.countInstance(
+      graph,
+      index,
+      retainedSizes,
+      excludeFromRetained: excludeFromRetained,
+    );
+    objects.add(index);
+    if (excludeFromRetained) objectsExcludedFromRetainedSize.add(index);
+  }
+
+  @override
+  void uncountInstance(
+    HeapSnapshotGraph graph,
+    int index,
+    List<int>? retainedSizes, {
+    required bool excludeFromRetained,
+  }) {
+    throw AssertionError('uncountInstance is not valid for $ObjectSet');
+  }
+}
+
+@immutable
+class ClassDataList<T extends ClassData> {
+  const ClassDataList(this._originalList)
+      : _appliedFilter = null,
+        _filtered = null;
+
+  const ClassDataList._filtered({
+    required List<T> original,
+    required ClassFilter appliedFilter,
+    required List<T> filtered,
+  })  : _originalList = original,
+        _appliedFilter = appliedFilter,
+        _filtered = filtered;
+
+  /// The list of classes after filtering.
+  List<T> get list => _filtered ?? _originalList;
+
+  final List<T> _originalList;
+  final ClassFilter? _appliedFilter;
+  final List<T>? _filtered;
+
+  Map<HeapClassName, T> asMap() =>
+      {for (var c in _originalList) c.className: c};
+
+  ClassDataList<T> filtered(ClassFilter newFilter, String? rootPackage) {
+    final filtered = ClassFilter.filter(
+      oldFilter: _appliedFilter,
+      oldFiltered: _filtered,
+      newFilter: newFilter,
+      original: _originalList,
+      extractClass: (s) => s.className,
+      rootPackage: rootPackage,
+    );
+    return ClassDataList._filtered(
+      original: _originalList,
+      appliedFilter: newFilter,
+      filtered: filtered,
+    );
+  }
+
+  T withMaxRetainedSize() {
+    return list.reduce(
+      (a, b) => a.objects.retainedSize > b.objects.retainedSize ? a : b,
+    );
+  }
+
+  /// Returns class data if [className] is presented in the [list].
+  ClassData? byName(HeapClassName? className) {
+    if (className == null) return null;
+    return list.firstWhereOrNull((c) => c.className == className);
+  }
+}
+
+abstract class ClassData {
+  ClassData({required this.className});
+
+  ObjectSetStats get objects;
+  Map<PathFromRoot, ObjectSetStats> get byPath;
+
+  final HeapClassName className;
+
+  bool contains(PathFromRoot? path) {
+    if (path == null) return false;
+    return byPath.containsKey(path);
+  }
+
+  late final PathFromRoot pathWithMaxRetainedSize = () {
+    assert(byPath.isNotEmpty);
+    return byPath.keys.reduce(
+      (a, b) => byPath[a]!.retainedSize > byPath[b]!.retainedSize ? a : b,
+    );
+  }();
+}
+
+class SingleClassData extends ClassData {
+  SingleClassData({required super.className});
+
+  @override
+  final ObjectSet objects = ObjectSet();
```

Comment:
nit: remove type on LHS

Reply:
In this case we are changing type from ObjectSetStats to more concrete ObjectSet. So it is needed.